### PR TITLE
Fix exception handling in /dart-internal-subscription to throw internal server error

### PR DESCRIPTION
**Pull request by ricardoamador**
The subscription handler runs its request zoned and as a result if the error returned is not an http exception/error the request will not be handled correctly. We will also be leaving a write lock in the cache.

*List which issues are fixed by this PR. You must list at least one issue.*
Fixes https://github.com/flutter/flutter/issues/135991

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: app_dart/lib/src/request_handlers/dart_internal_subscription.dart [hunk 1]
```diff
@@ -83,24 +84,31 @@ class DartInternalSubscription extends SubscriptionHandler {
...
-    } else {
-      log.info('Creating Task from Buildbucket result');
-      taskToInsert = await Task.fromBuildbucketBuild(build, datastore);
-    }
-
-    log.info('Inserting Task into the datastore: ${taskToInsert.toString()}');
-    await datastore.insert(<Task>[taskToInsert]);
 
-    return Body.forJson(taskToInsert.toString());
+    try {
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/dart_internal_subscription.dart:88]: Based on the backend log:
```
Uncaught error in request handler: NoSuchMethodError: Class 'KeyNotFoundException' has no instance getter 'statusCode'.
Receiver: Instance of 'KeyNotFoundException'
Tried calling: statusCode
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:38:5)
#1      SubscriptionHandler.service.<anonymous closure>.<anonymous closure> (package:cocoon_service/src/request_handling/subscription_handler.dart:142:65)
#2      RequestHandler.service.<anonymous closure> (package:cocoon_service/src/request_handling/request_handler.dart:69:26)
<asynchronous suspension>
#3      SubscriptionHandler.service (package:cocoon_service/src/request_handling/subscription_handler.dart:138:5)
<asynchronous suspension>
#4      main.<anonymous closure>.<anonymous closure> (file:///app/bin/server.dart:290:11)
<asynchronous suspension>
```

https://github.com/flutter/cocoon/blob/main/app_dart/lib/src/request_handling/request_handler.dart#L69 runs on `HttpStatusException`, which should have `statusCode` though?

Not sure if the try logic helps. With the try logic, I am wondering if the specified `InternalServerError` has any difference from the potential exception (from getbuild).

Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/dart_internal_subscription.dart:88]: I believe that is the assumption. Thought what happens when the exception thrown is not an HttpStatusException and not an error (which appears to be a different class of errors than exception) as this https://github.com/flutter/cocoon/blob/main/app_dart/lib/src/request_handling/request_handler.dart#L61 is the first pass at exception catching while the one you linked should be the catch all for whatever comes from those? 

If the exception is uncaught that means it is neither an HttpStatusException or an Error as is the case of the KeyNotFoundException which comes from the datastore.

Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/dart_internal_subscription.dart:88]: Though it is running the onError clause...

Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/dart_internal_subscription.dart:88]: I can duplicate this situation thought I am finding it hard to get the debugger to cooperate.
image

Comment (keyonghan) [reply app_dart/lib/src/request_handlers/dart_internal_subscription.dart:88]: Taking a deeper look, it is due to commit not found. 
https://github.com/flutter/cocoon/blob/main/app_dart/lib/src/request_handlers/dart_internal_subscription.dart#L98
https://github.com/flutter/cocoon/blob/main/app_dart/lib/src/model/appengine/task.dart#L161

With this patch, the API will throw internal error any way. I am wondering if we should consider fixing from datastore side. Like why the commit doesn't exist; Is it expected to exist?

Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/dart_internal_subscription.dart:88]: Yeah I am seeking to catch any non HttpRequestException here. Not just the keyNotFound exception. I don't think the datastore is wrong here in the implementation. It just means that whatever we are sending it is not in the datastore and from fromBuildBucketBuild we do not catch any exceptions, which is probably where we should actually fix this.

Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/dart_internal_subscription.dart:88]: I could catch datastore exceptions there and return a nullable Task which we could process to an internal server error.

Comment (keyonghan) [reply app_dart/lib/src/request_handlers/dart_internal_subscription.dart:88]: Do do still recall https://github.com/flutter/cocoon/pull/3005? Which seems expected to insert the branch commit.

Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/dart_internal_subscription.dart:88]: Yes, but it does a check to see if the commit already exists. Which is where it is failing. But now I see what you are talking about. When this is called it is assumed that the commit should exist. Okay let me investigate more.

**Comment by ricardoamador** [comment]
Reassigning to release team.

**Pull request closed without merging**